File: grpc-no-comments/test/distrib/bazel/cpp/greeter_server.cc

/*
 *
 * Copyright 2022 The gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/ext/proto_server_reflection_plugin.h>
#include <grpcpp/grpcpp.h>
#include <grpcpp/health_check_service_interface.h>

#include "protos/helloworld.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
using helloworld::Greeter;
using helloworld::HelloReply;
using helloworld::HelloRequest;

class GreeterServiceImpl final : public Greeter::Service {
  Status SayHello(ServerContext* context, const HelloRequest* request,
                  HelloReply* reply) override {
    std::string prefix("Hello ");
    reply->set_message(prefix + request->name());
    return Status::OK;
  }
};

void RunServer() {
  std::string server_address("0.0.0.0:0");
  GreeterServiceImpl service;

  grpc::EnableDefaultHealthCheckService(true);
  grpc::reflection::InitProtoReflectionServerBuilderPlugin();
  ServerBuilder builder;

  int bound_port;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials(), &bound_port);

  builder.RegisterService(&service);

  std::unique_ptr<Server> server(builder.BuildAndStart());
  std::cout << "127.0.0.1:" << bound_port << std::endl;

  server->Wait();
}

int main(int argc, char** argv) {
  RunServer();

  return 0;
}

File: grpc-result-comments-python/examples/python/interceptors/headers/greeter_server.py
Here's the commented version of the code:

```python
# Copyright 2017 gRPC authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""The Python implementation of the GRPC helloworld.Greeter server."""

# Import required modules
from concurrent import futures  # For thread pool execution
import logging  # For logging functionality

import grpc  # gRPC core library
import helloworld_pb2  # Generated protocol buffer messages
import helloworld_pb2_grpc  # Generated gRPC service stubs
from request_header_validator_interceptor import (
    RequestHeaderValidatorInterceptor,  # Custom interceptor for header validation
)

# Greeter service implementation class
class Greeter(helloworld_pb2_grpc.GreeterServicer):
    """Implementation of the Greeter service defined in the proto file."""
    
    def SayHello(self, request, context):
        """Implementation of the SayHello RPC method.
        
        Args:
            request: The HelloRequest message containing the 'name' field.
            context: The gRPC context for the RPC call.
            
        Returns:
            A HelloReply message with a greeting containing the request's name.
        """
        return helloworld_pb2.HelloReply(message="Hello, %s!" % request.name)

def serve():
    """Starts the gRPC server with request validation interceptor.
    
    The server:
    1. Sets up a header validator interceptor for authentication
    2. Creates a thread pool for handling requests
    3. Registers the Greeter service implementation
    4. Starts listening on port 50051
    """
    # Create header validator interceptor that checks for specific header
    header_validator = RequestHeaderValidatorInterceptor(
        "one-time-password",  # Required header key
        "42",  # Required header value
        grpc.StatusCode.UNAUTHENTICATED,  # Error code if validation fails
        "Access denied!",  # Error message if validation fails
    )
    
    # Create gRPC server with thread pool and interceptor
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),  # 10 worker threads
        interceptors=(header_validator,),  # Add our header validator interceptor
    )
    
    # Register our Greeter service implementation with the server
    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
    
    # Bind the server to port 50051 on all available interfaces
    server.add_insecure_port("[::]:50051")
    
    # Start the server
    server.start()
    
    # Keep the server running until terminated
    server.wait_for_termination()

if __name__ == "__main__":
    # Configure basic logging
    logging.basicConfig()
    
    # Start the server
    serve()
```